### PR TITLE
Fix MultiTrainer memory retention

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Summary
- run default MultiTrainer per-dataset training in CLI subprocesses and reload metrics from checkpoints
- stop retaining full trainer/model/dataloader objects in the parent process
- keep custom Python trainers/callbacks on the in-process path to preserve the public API
- reuse the package CLI launch prefix in Tuner without adding a public cfg API
- preserve in-memory metrics for no-checkpoint custom trainer runs and unique keys for repeated-dataset failures

## Tests
- python -m pytest tests/test_python.py::test_train_multi tests/test_engine.py::test_train_multi_custom_trainer_metrics_and_failure_keys -q
- python -m ruff check ultralytics/engine/model.py ultralytics/engine/trainer.py ultralytics/engine/tuner.py ultralytics/cfg/__init__.py tests/test_engine.py
- python smoke: YOLO("yolo26n.pt").train(data=["coco8.yaml", "coco128.yaml"], epochs=1, imgsz=32, batch=8, workers=0, plots=False, device="cpu") returned subprocess_path=True and fitness metrics for both datasets
- git diff --check

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
🚀 This PR improves multi-dataset training by making `MultiTrainer` more reliable across both Python and CLI workflows, while preserving metrics correctly and avoiding result-name conflicts.

### 📊 Key Changes
- 🧠 Updated multi-dataset training in `model.train()` to better decide when to use an in-process Python trainer versus launching CLI subprocesses.
- 🔁 Expanded `MultiTrainer` to support two execution modes:
  - direct Python trainer classes
  - CLI subprocess training when no custom trainer/callback setup is needed
- 💾 Added temporary base-model checkpoint handling for CLI-based multi-dataset runs so each dataset starts from the same model state.
- 📈 Improved metric collection logic:
  - keeps in-memory validator metrics when available
  - falls back to loading metrics from saved checkpoints if needed
- 🏷️ Fixed repeated-dataset naming so duplicate datasets get unique result keys like `coco8` and `coco8-2` instead of overwriting each other.
- 🛡️ Kept failure isolation in place: if one dataset run fails, the rest of the sweep continues, and the failed run is recorded as `None`.
- 🧪 Added a new test to confirm:
  - memory-resident metrics are preserved
  - duplicate dataset names stay unique
  - failed runs do not break the full multi-dataset training job
- 🧹 Refactored CLI launch command creation into a shared `_YOLO_CLI_COMMAND` constant, which is now reused in both trainer and tuner code.

### 🎯 Purpose & Impact
- ✅ Makes multi-dataset training more robust and predictable, especially when the same dataset appears more than once.
- 📊 Prevents metrics from being lost or overwritten, improving result tracking and comparison across runs.
- 🔄 Enables cleaner support for both advanced Python-based customization and standard CLI-style execution.
- 🚫 Reduces the chance that one failed dataset ruins an entire training sweep.
- 🧰 Simplifies internal command launching and improves maintainability by centralizing the CLI invocation logic.
- 👥 For users, this means more dependable multi-dataset fine-tuning with clearer outputs and better failure handling.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness—an engineer will assist soon. Please review the checklist below to help ensure smooth integration 😊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/engine/trainer.py [hunk 1]
```diff
@@ -1179,35 +1183,56 @@ def train(self):
...
+                            _get_yolo_cli_command("train", {k: v for k, v in overrides.items() if k != "session"}),
+                            check=True,
+                        )
+                    else:
+                        trainer = self.trainer(overrides=overrides, _callbacks=self.callbacks)
+                        trainer.model = trainer.get_model(weights=self.model, cfg=self.model.yaml)
+                        trainer.train()
+                    best, last = save_dir / "weights" / "best.pt", save_dir / "weights" / "last.pt"
+                    ckpt = best if best.exists() else last
+                    self.metrics[save_dir.name] = torch_load(ckpt)["train_metrics"] if ckpt.exists() else None
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py]: ⚠️ **HIGH**: This now relies exclusively on a checkpoint being written to populate `self.metrics`. In the in-process path, `save=False` (or any trainer that doesn't emit `best.pt`/`last.pt`) will make a successful run report `None`, whereas the previous implementation returned `trainer.validator.metrics.results_dict` from memory. Please keep the in-memory fallback for non-subprocess runs, otherwise multi-dataset training regresses for valid `save=False` usages.

File: ultralytics/engine/trainer.py [hunk 2]
```diff
@@ -1179,35 +1183,56 @@ def train(self):
...
+                    self.trainers.append(SimpleNamespace(save_dir=save_dir, best=best, last=last))
+                except Exception as e:  # one bad dataset should not abort the whole sweep
+                    LOGGER.error(f"MultiTrainer: fine-tuning on {data} failed, skipping: {e}")
+                    self.metrics[name] = None
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py]: 💡 **MEDIUM**: Failure cases are keyed by the raw dataset stem, while successful runs are keyed by the uniquified `save_dir.name`. With repeated datasets, a later failure can overwrite an earlier run's metrics entry (for example, the second `coco8.yaml` failing will replace the first run's successful `coco8` metrics with `None`). Use the resolved per-run name here as well so success and failure paths stay consistent.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The subprocess-based isolation is a solid approach for fixing trainer retention, but there are two regressions to address: metrics collection now breaks for successful runs that don't leave a checkpoint behind, and repeated-dataset failure bookkeeping can clobber an earlier run's metrics entry.

💬 Posted 2 inline comments

**Comment by glenn-jocher** [comment]
@fcakyon should fix memory leak by using Tuner strategy of subprocess training

**Comment by UltralyticsAssistant** [comment]
Merged — fantastic work, @glenn-jocher! 🚀

“Perfection is achieved, not when there is nothing more to add, but when there is nothing left to take away.” — Antoine de Saint-Exupéry

This update is a great example of that idea in practice: cleaner multi-dataset training logic, more reliable Python and CLI behavior, preserved metrics, and safer handling for duplicate datasets and failed runs. The result is a more robust and predictable experience for everyone training at scale.

Thanks for continuing to sharpen the core training workflow and making Ultralytics more dependable for the community.
